File: advent_of_code/2024/python/15.py
from typing import NamedTuple, Self


class Point(NamedTuple):
    x: int
    y: int

    def __add__(self, other: Self) -> Self:
        return Point(
            self.x + other.x,
            self.y + other.y,
        )


BOX = 'O'
ROBOT = '@'
SPACE = '.'

DIRECTIONS = {
    '<': Point(0, -1),
    '>': Point(0, 1),
    '^': Point(-1, 0),
    'v': Point(1, 0),
}


def within_bounds(p, grid):
    return 0 <= p.x < len(grid) and 0 <= p.y < len(grid[0])


def gps_coords(grid):
    total = 0
    for x in range(len(grid)):
        for y in range(len(grid[0])):
            if grid[x][y] == BOX:
                total += 100 * x + y
    return total


def do_move(d, grid):
    r = next(
        Point(x, y)
        for x in range(len(grid))
        for y in range(len(grid[0]))
        if grid[x][y] == ROBOT
    )
    group = []
    while within_bounds(r, grid) and grid[r.x][r.y] in [ROBOT, BOX]:
        group.append(r)
        r += d
    if within_bounds(r, grid) and grid[r.x][r.y] == SPACE:
        for p in group[::-1]:
            grid[p.x + d.x][p.y + d.y] = grid[p.x][p.y]
        if group:
            grid[p.x][p.y] = SPACE


def simulate_moves(grid, moves):
    for move in moves.replace('\n', ''):
        do_move(DIRECTIONS[move], grid)


def parse_grid(text):
    return list(map(list, text.splitlines()))


def grid_to_text(grid):
    return '\n'.join(''.join(line) for line in grid)


def part1(text):
    tests1()
    grid, moves = text.split('\n\n')
    grid = parse_grid(grid)
    simulate_moves(grid, moves.replace('\n', ''))
    return gps_coords(grid)


def part2(text):
    pass


def tests1():
    grid = parse_grid(
        '''#######
#...O..
#......'''
    )
    assert gps_coords(grid) == 104
    grid = parse_grid(
        '''########
#..O.O.#
##@.O..#
#...O..#
#.#.O..#
#...O..#
#......#
########'''
    )
    simulate_moves(grid, '<')
    assert (
        grid_to_text(grid)
        == '''########
#..O.O.#
##@.O..#
#...O..#
#.#.O..#
#...O..#
#......#
########'''
    )
    simulate_moves(grid, '^')
    assert (
        grid_to_text(grid)
        == '''########
#.@O.O.#
##..O..#
#...O..#
#.#.O..#
#...O..#
#......#
########'''
    )
    simulate_moves(grid, '^')
    assert (
        grid_to_text(grid)
        == '''########
#.@O.O.#
##..O..#
#...O..#
#.#.O..#
#...O..#
#......#
########'''
    )
    simulate_moves(grid, '>')
    assert (
        grid_to_text(grid)
        == '''########
#..@OO.#
##..O..#
#...O..#
#.#.O..#
#...O..#
#......#
########'''
    )
    simulate_moves(grid, '>')
    assert (
        grid_to_text(grid)
        == '''########
#...@OO#
##..O..#
#...O..#
#.#.O..#
#...O..#
#......#
########'''
    )
    simulate_moves(grid, '>')
    assert (
        grid_to_text(grid)
        == '''########
#...@OO#
##..O..#
#...O..#
#.#.O..#
#...O..#
#......#
########'''
    )
    simulate_moves(grid, 'v')
    assert (
        grid_to_text(grid)
        == '''########
#....OO#
##..@..#
#...O..#
#.#.O..#
#...O..#
#...O..#
########'''
    )
    simulate_moves(grid, 'v')
    assert (
        grid_to_text(grid)
        == '''########
#....OO#
##..@..#
#...O..#
#.#.O..#
#...O..#
#...O..#
########'''
    )
    simulate_moves(grid, '<')
    assert (
        grid_to_text(grid)
        == '''########
#....OO#
##.@...#
#...O..#
#.#.O..#
#...O..#
#...O..#
########'''
    )
    simulate_moves(grid, 'v')
    assert (
        grid_to_text(grid)
        == '''########
#....OO#
##.....#
#..@O..#
#.#.O..#
#...O..#
#...O..#
########'''
    )
    simulate_moves(grid, '>')
    assert (
        grid_to_text(grid)
        == '''########
#....OO#
##.....#
#...@O.#
#.#.O..#
#...O..#
#...O..#
########'''
    )
    simulate_moves(grid, '>')
    assert (
        grid_to_text(grid)
        == '''########
#....OO#
##.....#
#....@O#
#.#.O..#
#...O..#
#...O..#
########'''
    )
    simulate_moves(grid, 'v')
    assert (
        grid_to_text(grid)
        == '''########
#....OO#
##.....#
#.....O#
#.#.O@.#
#...O..#
#...O..#
########'''
    )
    simulate_moves(grid, '<')
    assert (
        grid_to_text(grid)
        == '''########
#....OO#
##.....#
#.....O#
#.#O@..#
#...O..#
#...O..#
########'''
    )
    simulate_moves(grid, '<')
    assert (
        grid_to_text(grid)
        == '''########
#....OO#
##.....#
#.....O#
#.#O@..#
#...O..#
#...O..#
########'''
    )
    assert gps_coords(grid) == 2028

File: advent_of_code/2024/rust/src/15.rs
use phf::phf_map;
use std::collections::HashSet;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Point {
    x: i32,
    y: i32,
}

impl Point {
    fn add(self, other: Self) -> Self {
        Point {
            x: self.x + other.x,
            y: self.y + other.y,
        }
    }
}

const BOX: char = 'O';
const LEFT_BOX: char = '[';
const RIGHT_BOX: char = ']';
const ROBOT: char = '@';
const SPACE: char = '.';
const WALL: char = '#';

const LEFT: Point = Point { x: 0, y: -1 };
const RIGHT: Point = Point { x: 0, y: 1 };
const UP: Point = Point { x: -1, y: 0 };
const DOWN: Point = Point { x: 1, y: 0 };
const DIRECTIONS: phf::Map<char, Point> = phf_map! {
    '<' => LEFT,
    '>' => RIGHT,
    '^' => UP,
    'v' => DOWN,
};

fn within_bounds(p: Point, grid: &Vec<Vec<char>>) -> bool {
    p.x >= 0 && p.x < grid.len() as i32 && p.y >= 0 && p.y < grid[0].len() as i32
}

fn gps_coords(grid: &Vec<Vec<char>>) -> i32 {
    let mut total = 0;
    for x in 0..grid.len() {
        for y in 0..grid[0].len() {
            if grid[x][y] == BOX || grid[x][y] == LEFT_BOX {
                total += 100 * x as i32 + y as i32;
            }
        }
    }
    total
}

fn find_push_group(r: Point, d: Point, grid: &Vec<Vec<char>>) -> HashSet<Point> {
    if !within_bounds(r, grid) {
        return HashSet::new();
    }

    let mut stack = vec![r];
    let mut group = HashSet::new();

    while let Some(p) = stack.pop() {
        if !group.contains(&p)
            && grid[p.x as usize][p.y as usize] != WALL
            && grid[p.x as usize][p.y as usize] != SPACE
        {
            group.insert(p);

            if matches!(
                grid[p.x as usize][p.y as usize],
                ROBOT | BOX | LEFT_BOX | RIGHT_BOX
            ) || (grid[p.x as usize][p.y as usize] == LEFT_BOX && d.x != 0)
            {
                let next_p = p.add(d);
                if !group.contains(&next_p) {
                    stack.push(next_p);
                }
            }

            if grid[p.x as usize][p.y as usize] == LEFT_BOX && d.x != 0 {
                let right_p = p.add(RIGHT);
                if !group.contains(&right_p) {
                    stack.push(right_p);
                }
            } else if grid[p.x as usize][p.y as usize] == RIGHT_BOX && d.x != 0 {
                let left_p = p.add(LEFT);
                if !group.contains(&left_p) {
                    stack.push(left_p);
                }
            }
        }
    }

    group
}

fn do_move(d: Point, grid: &mut Vec<Vec<char>>) {
    let robot_position = grid.iter().enumerate().find_map(|(x, row)| {
        row.iter().position(|&cell| cell == ROBOT).map(|y| Point {
            x: x as i32,
            y: y as i32,
        })
    });

    if let Some(r) = robot_position {
        let group = find_push_group(r, d, grid);
        let pushed_group: Vec<(Point, char)> = group
            .iter()
            .map(|&p| (p.add(d), grid[p.x as usize][p.y as usize]))
            .collect();

        if pushed_group
            .iter()
            .all(|&(p, _)| within_bounds(p, grid) && grid[p.x as usize][p.y as usize] != WALL)
        {
            for &p in &group {
                grid[p.x as usize][p.y as usize] = SPACE;
            }

            for &(p, v) in &pushed_group {
                grid[p.x as usize][p.y as usize] = v;
            }
        }
    }
}

fn simulate_moves(grid: &mut Vec<Vec<char>>, moves: &str) {
    for mv in moves.chars() {
        if let Some(&direction) = DIRECTIONS.get(&mv) {
            do_move(direction, grid);
        }
    }
}

fn parse_grid(text: &str) -> Vec<Vec<char>> {
    text.lines().map(|line| line.chars().collect()).collect()
}

fn part1(text: &str) -> i32 {
    let parts: Vec<&str> = text.split("\n\n").collect();
    let mut grid = parse_grid(parts[0]);
    simulate_moves(&mut grid, parts[1]);
    gps_coords(&grid)
}

fn part2(text: &str) -> i32 {
    let parts: Vec<&str> = text.split("\n\n").collect();
    let mut grid = parse_grid(
        &parts[0]
            .replace('#', "##")
            .replace('O', "[]")
            .replace('.', "..")
            .replace('@', "@."),
    );
    simulate_moves(&mut grid, parts[1]);
    gps_coords(&grid)
}

fn main() {
    aoc2024::solve(part1, part2);
}
